base/sort.jl

```diff
@@ -367,6 +409,40 @@ end
 
 sort(v::AbstractVector; kws...) = sort!(copy(v); kws...)
 
+
+## selectperm: the permutation to sort the first k elements of an array ##
+
+function selectperm(v::AbstractVector,
+                    k::Union(Int,OrdinalRange);
+                    lt::Function=isless,
+                    by::Function=identity,
+                    rev::Bool=false,
+                    order::Ordering=Base.Order.Forward)
+    select!(collect(1:length(v)), k, Perm(ord(lt, by, rev, order), v))
+end
+
+function selectperm!{I<:Integer}(ix::AbstractVector{I}, v::AbstractVector,
+                                 k::Union(Int, OrdinalRange);
+                                 lt::Function=isless,
+                                 by::Function=identity,
+                                 rev::Bool=false,
+                                 order::Ordering=Forward,
+                                 initialized::Bool=false)
+    if !initialized
+        @inbounds for i = 1:length(ix)
+            ix[i] = i
+        end
+    end
+
+    # do partial quicksort
+    sort!(ix, PartialQuickSort(k), Perm(ord(lt, by, rev, order), v))
+
+    # TODO: Not type stable. If k is an int, this will return an Int, of it is
+    #       an OrdinalRange it will return a Vector{Int}. This, however, seems
+    #       to be the same behavior as as `select`
```

Comment:
Actually, your description shows that [this function is actually type-stable](http://julia.readthedocs.org/en/latest/manual/faq/#what-does-type-stable-mean).